### ISSUE TITLE
Emit helpful `note` messages only once

### ISSUE THREAD

**Issue by bluss** [rust-lang/rust#24690]
Example: Adding `#![deny(lint_name)]` to a crate will point out the location of this line for every time the lint fires (`note: lint level defined here`). This should be listed only once. It should be reasonably easy to deduplicate these based upon equal span and message.

```
src/adaptors.rs:84:5: 87:6 warning: missing documentation for a method
src/adaptors.rs:84     pub fn new(iter: I, map: fn(I::Item) -> B) -> Self
src/adaptors.rs:85     {
src/adaptors.rs:86         FnMap{iter: iter, map: map}
src/adaptors.rs:87     }
src/lib.rs:1:9: 1:21 note: lint level defined here
src/lib.rs:1 #![warn(missing_docs)]
                     ^~~~~~~~~~~~
src/islice.rs:30:5: 37:6 warning: missing documentation for a method
src/islice.rs:30     pub fn new<R: GenericRange>(iter: I, range: R) -> Self
src/islice.rs:31     {
src/islice.rs:32         ISlice {
src/islice.rs:33             start: range.start().unwrap_or(0),
src/islice.rs:34             end: range.end().unwrap_or(::std::usize::MAX),
src/islice.rs:35             iter: iter,
                 ...
src/lib.rs:1:9: 1:21 note: lint level defined here
src/lib.rs:1 #![warn(missing_docs)]
                     ^~~~~~~~~~~~
src/misc.rs:15:5: 15:17 warning: missing documentation for an associated type
src/misc.rs:15     type Result;
                   ^~~~~~~~~~~~
src/lib.rs:1:9: 1:21 note: lint level defined here
src/lib.rs:1 #![warn(missing_docs)]
```

**Comment by mitaa** [comment]
I think this is a good "E-easy" candidate.

Relevant code is in [src/librustc/lint/context.rs](https://github.com/rust-lang/rust/blob/cee9463d24ee3e5682888aa4ddfbcfc01a1f5e83/src/librustc/lint/context.rs#L470).
Using a `HashSet<(Span, &str)>` as a field on `LintStore` should work, so that we only emit a note the first time we encounter it.

**Comment by nikomatsakis** [comment]
Question -- in #37191, we get only "lint level defined here" per `#[allow]`, even if it applies to multiple lints (e.g., `#[allow(warnings)]`). Do people think it'd be better to have it like this, or would you rather see different messages for (say) `dead_code` and `unused_variable`, even if they both apply to the same `#[allow]`?

**Comment by zackmdavis** [comment]
Can someone with perms close this? I forgot to use the [magic words](https://github.com/blog/1506-closing-issues-via-pull-requests) in #37191.

**Comment by bluss** [comment]
Great! Fixed in #37191
